File: week 4/majority_element/majority_element.py
# Uses python3
import sys

def get_majority_element_on(a):
    map_by_count = {}

    for i in range(len(a)):
        if a[i] in map_by_count:
            map_by_count[a[i]] += 1
        else:
            map_by_count[a[i]] = 1

    threshold = len(a) // 2 + 1
    for (key, val) in map_by_count.items():
        if val >= threshold:
            return 1

    return 0

def get_majority_element(a, left, right):
    if left + 1 == right:
        return

    mid = (left + right) // 2
    get_majority_element(a, left, mid)
    get_majority_element(a, mid, right)

    first = a[left:mid]

    i = 0
    j = mid
    k = left

    while i + left < mid and j < right:
        if first[i] <= a[j]:
            a[k] = first[i]
            i += 1
        else:
            a[k] = a[j]
            j += 1
        k += 1

    while left + i < mid:
        a[k] = first[i]
        k += 1
        i += 1

    if left == 0 and right == len(a):
        if a[mid] == a[0] or a[mid - 1] == a[right - 1]:
            return 1
        else:
            return 0

def test_solution():
    assert get_majority_element_on([1, 2, 3, 4]) == 0
    assert get_majority_element_on([2, 2, 3, 4]) == 0
    assert get_majority_element_on([2, 3, 2, 3, 2, 2, 3, 2]) == 1
    assert get_majority_element([1, 2, 3, 4], 0, 4) == 0
    assert get_majority_element([2, 2, 3, 4], 0, 4) == 0
    assert get_majority_element([2, 2, 3, 3, 2, 2, 2, 3], 0, 8) == 1

if __name__ == '__main__':
    test_solution()

    input = sys.stdin.read()
    n, *a = list(map(int, input.split()))
    if get_majority_element(a, 0, n) != -1:
        print(1)
    else:
        print(0)

File: week 4/majority_element/majority_element.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <cassert>

using std::vector;
using std::cout;

int get_majority_element(vector<int> &a, int left, int right) {
  if (left + 1 == right) {
    return 0;
  }

  int mid = (left + right) / 2;
  get_majority_element(a, left, mid);
  get_majority_element(a, mid, right);

  std::vector<int> first{mid - left - 1};
  for (int b = left; b < mid; b++) {
    first[b - left] = a[b];
  }

  int i = 0, j = mid, k = left;
  while (i + left < mid && j < right) {
    if (first[i] <= a[j]) {
      a[k++] = first[i++];
    } else {
      a[k++] = a[j++];
    }
  }

  while (left + i < mid) {
    a[k++] = first[i++];
  }

  if (left == 0 && right == a.size()) {
    return (a[mid] == a[0] || a[mid - 1] == a[right - 1]);
  }

  return 0;
}

void test_solution() {
  std::vector<int> v;
  v = {1, 2, 3, 4};
  assert(get_majority_element(v, 0, 4) == 0);
  v = {2, 2, 3, 4, 5};
  assert(get_majority_element(v, 0, 5) == 0);
  v = {2, 2, 3, 3, 2, 2, 2, 3};
  assert(get_majority_element(v, 0, 8) == 1);
}

int main() {
  test_solution();

  int n;
  std::cin >> n;
  vector<int> a(n);
  for (size_t i = 0; i < a.size(); ++i) {
    std::cin >> a[i];
  }
  std::cout << (get_majority_element(a, 0, a.size()) != -1) << '\n';
}
